packages/dataviews/src/hooks/use-form-validity.ts

```diff
@@ -74,486 +76,617 @@ function isFormValid( formValidity: FormValidity | undefined ): boolean {
-						return prev;
-					}
-
-					const { [ field.id ]: removed, ...restChildren } =
-						parentField.children as any;
-
-					// If no more children, remove the children property
-					if ( Object.keys( restChildren ).length === 0 ) {
-						const { children, ...restParent } = parentField;
-						if ( Object.keys( restParent ).length === 0 ) {
-							// Remove parent field entirely if no other validations
-							const {
-								[ parentFieldId ]: removedParent,
-								...restFields
-							} = prev;
-							return Object.keys( restFields ).length === 0
-								? undefined
-								: restFields;
-						}
-						return {
-							...prev,
-							[ parentFieldId ]: restParent,
-						};
-					}
-
-					return {
-						...prev,
-						[ parentFieldId ]: {
-							...parentField,
-							children: restChildren,
-						},
-					};
-				}
+	// Validate the field: isValid.custom (async)
+	if ( customError instanceof Promise ) {
+		handleCustomValidationAsync( customError, formField, promiseHandler );
 
-				// Regular field - remove from top level
-				if ( ! prev[ field.id ] ) {
-					return prev;
-				}
+		return {
+			custom: {
+				type: 'validating',
+				message: __( 'Validating…' ),
+			},
+		};
+	}
 
-				const { [ field.id ]: removed, ...rest } = prev;
+	// Validate its children.
+	if ( formField.children ) {
+		const result: Record< string, FieldValidity | undefined > = {};
+		formField.children.forEach( ( child ) => {
+			result[ child.id ] = validateFormField( item, child, {
+				...promiseHandler,
+				path: [ ...promiseHandler.path, formField.id, 'children' ],
+			} );
+		} );
 
-				if ( Object.keys( rest ).length === 0 ) {
-					return undefined;
-				}
+		const filteredResult: Record< string, FieldValidity > = {};
+		Object.entries( result ).forEach( ( [ key, value ] ) => {
+			if ( value !== undefined ) {
+				filteredResult[ key ] = value;
+			}
+		} );
+
+		if ( Object.keys( filteredResult ).length === 0 ) {
+			return undefined;
+		}
+
+		return {
+			children: filteredResult,
+		};
+	}
+
+	// No errors for this field or its children.
+	return undefined;
+}
 
-				return rest;
+function getFormFieldValue< Item >(
+	formField: FormFieldToValidate< Item >,
+	item: Item
+): any {
+	const fieldValue = formField?.field?.getValue( { item } );
+	if ( ! formField.children ) {
+		return fieldValue;
+	}
+
+	const childrenValues =
+		formField.children.length > 0
+			? formField.children.map( ( child ) =>
+					getFormFieldValue( child, item )
+			  )
+			: undefined;
+
+	if ( ! childrenValues ) {
+		return fieldValue;
+	}
+
+	return {
+		value: fieldValue,
+		children: childrenValues,
+	};
+}
+
+/**
+ * Hook that validates a form item and returns an object with error messages for each field.
+ *
+ * @param item   The item to validate.
+ * @param fields Fields config.
+ * @param form   Form config.
+ *
+ * @return Record of field IDs to error messages (undefined means no error).
+ */
+export function useFormValidity< Item >(
```

Comment:
This function has been completely refactored, so the diff would not be super helpful to understand what's actually changed. For review, I recommend just taking a look at how it works now and ignore the previous version.

Reply:
I think this PR would have been a good opportunity to refactor to delegate the validation behavior to the field types instead of centralizing it here.

Reply:
This PR updates the core flow to fix the existing issue. Extracting the existing logic to the field type definitions is an extra thing on top of what this PR already does. The PR is already big as it is, so I thought I'd control scope by focusing just on fixing the issue.